transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.pool;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.EventLoop;
+import io.netty.util.AttributeKey;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.ImmediateEventExecutor;
+import io.netty.util.concurrent.Promise;
+import io.netty.util.internal.PlatformDependent;
+
+import java.util.Deque;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * Simple {@link ChannelPool} implementation which will create new {@link Channel}s if someone tries to acquire
+ * a {@link Channel} but none is in the pool atm. No limit on the maximal concurrent {@link Channel}s is enforced.
+ *
+ * @param <C>   the {@link Channel} type to pool.
+ * @param <K>   the {@link ChannelPoolKey} that is used to store and lookup the {@link Channel}s.
+ */
+public final class SimpleChannelPool<C extends Channel, K extends ChannelPoolKey> implements ChannelPool<C, K> {
+
+    private static final AttributeKey<ChannelPoolKey> KEY = AttributeKey.newInstance("channelPoolKey");
+    private final ConcurrentMap<ChannelPoolKey, Deque<C>> pool = PlatformDependent.newConcurrentHashMap();
+    private final ChannelPoolHandler<C, K> handler;
+    private final ChannelPoolHealthChecker<C, K> healthCheck;
+    private final Bootstrap bootstrap;
+
+    public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler<C, K> handler) {
```

Comment:
@pjulien The problem is that adding handlers in the future may be too late as at this point the channel is connected already and you may need to handle for example registered and active events in your ChannelHandler. This is the same reason why we allow to set a handler on the Bootstrap itself when doing normal bootstrapping.